src/Microsoft.SqlTools.ServiceLayer/Agent/AgentService.cs

```diff
@@ -148,183 +148,174 @@ public void InitializeService(ServiceHost serviceHost)
-                        int count = dt.Rows.Count;
-                        List<AgentJobHistoryInfo> jobHistories = new List<AgentJobHistoryInfo>();
-                        if (count > 0)
-                        {
-                            var job = dt.Rows[0];
-                            Guid jobId = (Guid)job[AgentUtilities.UrnJobId];
-                            int runStatus = Convert.ToInt32(job[AgentUtilities.UrnRunStatus], System.Globalization.CultureInfo.InvariantCulture);
-                            var t = new LogSourceJobHistory(parameters.JobName, sqlConnInfo, null, runStatus, jobId, null);
-                            var tlog = t as ILogSource;
-                            tlog.Initialize();
-                            var logEntries = t.LogEntries;
-
-                            // Finally add the job histories
-                            jobHistories = AgentUtilities.ConvertToAgentJobHistoryInfo(logEntries, job, steps);
-                            result.Histories = jobHistories.ToArray();
-                            result.Success = true;
-                            tlog.CloseReader();
-                        }
-                        await requestContext.SendResult(result);
+                    // Add histories
+                    int count = dt.Rows.Count;
+                    List<AgentJobHistoryInfo> jobHistories = new List<AgentJobHistoryInfo>();
+                    if (count > 0)
+                    {
+                        var job = dt.Rows[0];
+                        Guid jobId = (Guid)job[AgentUtilities.UrnJobId];
+                        int runStatus = Convert.ToInt32(job[AgentUtilities.UrnRunStatus], System.Globalization.CultureInfo.InvariantCulture);
+                        var t = new LogSourceJobHistory(parameters.JobName, sqlConnInfo, null, runStatus, jobId, null);
+                        var tlog = t as ILogSource;
+                        tlog.Initialize();
+                        var logEntries = t.LogEntries;
+
+                        // Finally add the job histories
+                        jobHistories = AgentUtilities.ConvertToAgentJobHistoryInfo(logEntries, job, steps);
+                        result.Histories = jobHistories.ToArray();
+                        result.Success = true;
+                        tlog.CloseReader();
                     }
+                    await requestContext.SendResult(result);
                 }
-                catch (Exception e)
-                {
-                    await requestContext.SendError(e);
-                }
-            });
+            }
+            catch (Exception e)
+            {
+                await requestContext.SendError(e);
+            }
         }
 
         /// <summary>
         /// Handle request to Run a Job
         /// </summary>
         internal async Task HandleJobActionRequest(AgentJobActionParams parameters, RequestContext<ResultStatus> requestContext)
         {
-            await Task.Run(async () =>
```

Comment:
Sorry it's taken me long to get back to this one. Super busy, and only just now got some free time back. I'm not entirely sure about whether this changes the order of execution, especially the comment about CPU scheduler vs task scheduler, since the method is still async and returns a Task. Either way, I'd be fine with taking your suggestion to return the T`ask.Run(async ()`.

Reply:
So reading more into this it seems I wasn't quite correct - this would only be a problem in situations where a particular SynchronizationContext is used that would cause the blocking - e.g. a GUI or ASP.NET application. For console apps like this though it'll happily run the awaited code on a different thread. 

See https://docs.microsoft.com/en-us/archive/msdn-magazine/2013/march/async-await-best-practices-in-asynchronous-programming#async-all-the-way (particularly after the deadlock demo)

Here's an example program demonstrating that : https://gist.github.com/Charles-Gagnon/f8e3f4c7c115b9f99d3c3f2d0f8234bd

Output : 

```
Main Thread=1
Doing work for Task2... Thread=1
Doing work for Task1... Thread=9
Sleeping...
Doing work for Task2... Thread=9
Doing work for Task1... Thread=6
Doing work for Task2... Thread=9
Doing work for Task1... Thread=12
Doing work for Task2... Thread=9
Doing work for Task1... Thread=11
Doing work for Task2... Thread=9
Doing work for Task1... Thread=6
Doing work for Task2... Thread=6
Doing work for Task1... Thread=9
Doing work for Task1... Thread=6
Doing work for Task2... Thread=9
Doing work for Task2... Thread=6
Doing work for Task1... Thread=9
Doing work for Task2... Thread=6
Doing work for Task1... Thread=11
Doing work for Task1... Thread=6
Doing work for Task2... Thread=11
```

Notice how it'll start off using the same thread but as soon as it hits the await then the main thread will continue, and since it hits the Sleep when the Task wakes back up it'll just happily continue on a different thread.

Reply:
Oh sweet! Thanks for validating that behavior! I'll definitely take a good look at the links you provided to enhance my understanding. Thanks for merging this!